File: best_time_buy_sell_ii/max_profit_ii.go
package besttimebuysellii

func maxProfit(prices []int) int {
	if len(prices) == 1 {
		return 0
	}
	buyPrice := prices[0]
	maxProfit := 0

	for i := 1; i < len(prices); i++ {
		if prices[i] < buyPrice {
			buyPrice = prices[i]
		} else {
			profit := prices[i] - buyPrice

			if profit > 0 {
				maxProfit += profit
				buyPrice = prices[i]
			}
		}
	}

	return maxProfit
}

File: best_time_buy_sell_ii/max_profit_ii.dart
class Solution {
  int maxProfit(List<int> prices) {
    if (prices.length == 1) {
      return 0;
    }

    var buyPrice = prices[0];
    var maxProfit = 0;

    for (var i = 1; i < prices.length; i++) {
      if (prices[i] < buyPrice) {
        buyPrice = prices[i];
      } else {
        var profit = prices[i] - buyPrice;

        if (profit > 0) {
          maxProfit += profit;
          buyPrice = prices[i];
        }
      }
    }
    return maxProfit;
  }
}
